lib/engine-object-file/README.md

```diff
@@ -0,0 +1,186 @@
+        current_offset += sizeof(size_t);
+
+        memcpy(memory_buffer + current_offset, (void*)&function_trampolines[0], sizeof(function_trampolines));
+        current_offset += sizeof(function_trampolines);
+
+        memcpy(memory_buffer + current_offset, (void*)&num_dynamic_function_trampoline_pointers, sizeof(size_t));
+        current_offset += sizeof(size_t);
+
+        memcpy(memory_buffer + current_offset, (void*)&dynamic_function_trampoline_pointers[0], sizeof(dynamic_function_trampoline_pointers));
+        current_offset += sizeof(dynamic_function_trampoline_pointers);
+
+        my_byte_vec_t module_byte_vec = {
+                .size = buffer_size,
+                .data = memory_buffer,
+        };
+
+        wasm_module_t* module = wasm_module_deserialize(store, (wasm_byte_vec_t*) &module_byte_vec);
+        if (! module) {
+                printf("Failed to create module\n");
+                print_wasmer_error();
+                return -1;
+        }
+        free(memory_buffer);
+        
+        // We have now finished the memory buffer book keeping and we have a valid Module.
+
+        // In this example we're passing some JavaScript source code as a command line argumnet
+        // to a WASI module that can evaluate JavaScript.
+        wasi_config_t* config = wasi_config_new("constant_value_here");
+        const char* js_string = "function greet(name) { return JSON.stringify('Hello, ' + name); }; print(greet('World'));";
+        wasi_config_arg(config, "--eval");
+        wasi_config_arg(config, js_string);
+        wasi_env_t* wasi_env = wasi_env_new(config);
+        if (!wasi_env) {
+                printf("> Error building WASI env!\n");
+                print_wasmer_error();
+                return 1;
+        }
+
+        wasm_importtype_vec_t import_types;
+        wasm_module_imports(module, &import_types);
+        int num_imports = import_types.size;
+        wasm_extern_t** imports = malloc(num_imports * sizeof(wasm_extern_t*));
+        wasm_importtype_vec_delete(&import_types);
+        
+        bool get_imports_result = wasi_get_imports(store, module, wasi_env, imports);
+        if (!get_imports_result) {
+                printf("> Error getting WASI imports!\n");
+                print_wasmer_error();
+                return 1;
+        }
+
+        wasm_instance_t* instance = wasm_instance_new(store, module, (const wasm_extern_t* const*) imports, NULL);
+        if (! instance) {
+                printf("Failed to create instance\n");
+                print_wasmer_error();
+                return -1;
+        }
+        wasi_env_set_instance(wasi_env, instance);
+        
+        // WASI is now set up.
+
+        void* vmctx = wasm_instance_get_vmctx_ptr(instance);
+        wasm_val_t* inout[2] = { NULL, NULL };
+
+        // We're able to call our compiled functions directly through their trampolines.
+        wasmer_trampoline_function_call__1(vmctx, wasmer_function__1, &inout);
```

Comment:
We should probably just use the normal wasm c api for calling functions.
Probably this code is not needed?

Reply:
There's a series of type requirements that we can't meet:

Quoting from when I wrote about it on Slack previously:

----

`wasm_func_callback_t` = `struct wasm_trap_t *(*)(const struct wasm_val_t *, struct wasm_val_t *)`.
`wasm_func_new` takes `wasm_func_callback_t` + `wasm_functype_t`  and returns `wasm_func_t`
`wasm_func_call` takes `wasm_func_t` + args + rets returns `wasm_trap_t*`

```
#[repr(C)]
pub struct wasm_func_t {
    inner: Function,
    // this is how we ensure the instance stays alive
    instance: Option<Arc<Instance>>,
}
```
Is our current definition of `wasm_func_t`

We don’t have anything of the correct type to use our functions as `wasm_func_t` right now
it may be possible to change `wasm_func_t` and/or make a `wasmer_func_new` or something to make things work
but it’s invalid to cast any of our pointers to a `wasm_func_callback_t` or `wasm_func_t` right now

Reply:
Following up on my previous comment, now that we have traps implemented... it's probably possible to generate wrapper functions so that we can have a bunch of `wasm_func_t`s... we'd need a helper function or two 🤔 ; fundamentally what we're doing is making a trampoline so that we can make another trampoline for it though... so perhaps having a wasmer-specific lower level method to get a `wasm_func_t` directly is the best option here